hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/cache/EntityCache.java

```diff
@@ -30,105 +30,102 @@
-          .maximumSize(10000)
-          .expireAfterWrite(5, TimeUnit.MINUTES)
-          .build(
-              new CacheLoader<>() {
-                @Override
-                public List<Entity> load(@Nonnull Pair<String, String> key) {
-                  return edsClient.getEntitiesByName(
-                      key.getLeft(), EntityType.K8S_NAMESPACE.name(), key.getRight());
-                }
-              });
+  private final LoadingCache<Pair<String, String>, List<Entity>> nameToNamespaceEntitiesCache;
 
   /**
    * Cache of Backend identifying attributes to Entity Key: Map of identifying attributes Value:
    * Optional Backend entity
    */
   private final LoadingCache<ContextualKey<Map<String, AttributeValue>>, Optional<Entity>>
-      backendIdAttrsToEntityCache =
-          CacheBuilder.newBuilder()
-              .maximumSize(10000)
-              .expireAfterWrite(5, TimeUnit.MINUTES)
-              .build(CacheLoader.from(this::loadBackendFromIdentifyingAttributes));
+      backendIdAttrsToEntityCache;
 
-  public EntityCache(EdsClient edsClient) {
+  public EntityCache(EdsClient edsClient, Executor asyncCacheLoaderExecutor) {
     this.edsClient = edsClient;
+    fqnToServiceEntityCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(10000)
+            .refreshAfterWrite(4, TimeUnit.MINUTES)
+            .expireAfterWrite(5, TimeUnit.MINUTES)
+            .recordStats()
+            .build(
+                CacheLoader.asyncReloading(
+                    CacheLoader.from(this::loadServiceFromFQN), asyncCacheLoaderExecutor));
+
+    nameToServiceEntitiesCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(10000)
+            .refreshAfterWrite(4, TimeUnit.MINUTES)
+            .expireAfterWrite(5, TimeUnit.MINUTES)
+            .build(
+                CacheLoader.asyncReloading(
+                    CacheLoader.from(
+                        tenantIdServiceNamePair ->
+                            edsClient.getEntitiesByName(
+                                tenantIdServiceNamePair.getLeft(),
+                                EntityType.SERVICE.name(),
+                                tenantIdServiceNamePair.getRight())),
+                    asyncCacheLoaderExecutor));
+
+    nameToNamespaceEntitiesCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(10000)
+            .refreshAfterWrite(4, TimeUnit.MINUTES)
+            .expireAfterWrite(5, TimeUnit.MINUTES)
+            .build(
+                CacheLoader.asyncReloading(
+                    CacheLoader.from(
+                        key ->
+                            edsClient.getEntitiesByName(
+                                key.getLeft(), EntityType.K8S_NAMESPACE.name(), key.getRight())),
+                    asyncCacheLoaderExecutor));
+
+    backendIdAttrsToEntityCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(10000)
+            .expireAfterWrite(5, TimeUnit.MINUTES)
```

Comment:
Are we missing `refreshAfterWrite` here ?

Reply:
Yes. Fixed it.